consensus/src/experimental/commit_phase.rs

```diff
@@ -0,0 +1,333 @@
+    commit_channel_recv: Receiver<(Vec<ExecutedBlock>, LedgerInfoWithSignatures)>,
+    execution_proxy: Arc<dyn StateComputer>,
+    local_cache: HashMap<HashValue, LedgerInfoWithSignatures>,
+    local_queue: VecDeque<CacheItem>,
+    commit_msg_sender: channel::Sender<CommitPhaseChannelMsgWrapper>,
+    commit_msg_receiver: channel::Receiver<ConsensusMsg>,
+    verifier: ValidatorVerifier,
+    safety_rules: Arc<Mutex<MetricsSafetyRules>>,
+    author: Author,
+    committed_round: Round,
+}
+
+pub async fn commit(
+    execution_proxy: &Arc<dyn StateComputer>,
+    vecblock: &[ExecutedBlock],
+    ledger_info: &LedgerInfoWithSignatures,
+) -> Result<(), ExecutionError> {
+    // have to maintain the order.
+    execution_proxy
+        .commit(
+            &vecblock
+                .iter()
+                .map(|eb| Arc::new(eb.clone()))
+                .collect::<Vec<Arc<ExecutedBlock>>>(),
+            ledger_info.clone(),
+        )
+        .await
+}
+
+impl CommitPhase {
+    pub fn new(
+        commit_channel_recv: Receiver<(Vec<ExecutedBlock>, LedgerInfoWithSignatures)>,
+        execution_proxy: Arc<dyn StateComputer>,
+        commit_msg_sender: channel::Sender<CommitPhaseChannelMsgWrapper>,
+        commit_msg_receiver: channel::Receiver<ConsensusMsg>,
+        verifier: ValidatorVerifier,
+        safety_rules: Arc<Mutex<MetricsSafetyRules>>,
+        author: Author,
+    ) -> Self {
+        Self {
+            commit_channel_recv,
+            execution_proxy,
+            local_cache: HashMap::<HashValue, LedgerInfoWithSignatures>::new(),
+            local_queue: VecDeque::<CacheItem>::new(),
+            commit_msg_sender,
+            commit_msg_receiver,
+            verifier,
+            safety_rules,
+            author,
+            committed_round: 0,
+        }
+    }
+
+    /// Notified when receiving a commit vote message
+    pub async fn process_commit_vote(&mut self, commit_vote: &CommitVote) -> anyhow::Result<()> {
+        let li = commit_vote.ledger_info();
+
+        if li.commit_info().round() < self.committed_round {
+            return Ok(()); // we ignore the message
+        }
+
+        // verify the signature
+        commit_vote.verify(&self.verifier)?;
+
+        let chash = li.commit_info().executed_state_id();
+
+        // add the signature to local_cache
+        match self.local_cache.entry(chash) {
+            Entry::Occupied(mut cio) => {
+                let ci = cio.get_mut();
+                ci.add_signature(commit_vote.author(), commit_vote.signature().clone());
+            }
+            Entry::Vacant(_) => {
+                let mut li_sig = LedgerInfoWithSignatures::new(
+                    li.clone(),
+                    BTreeMap::<AccountAddress, Ed25519Signature>::new(),
+                );
+                li_sig.add_signature(commit_vote.author(), commit_vote.signature().clone());
+                self.local_cache.insert(chash, li_sig);
+            }
+        };
```

Comment:
this can be simplified with  map.entry(hash).or_insert_with(||..).add_signature(..)